test/Npgsql.Tests/BugTests.cs

```diff
@@ -1381,5 +1383,50 @@ public async Task Bug3924()
                 Assert.DoesNotThrowAsync(cmd.ExecuteNonQueryAsync);
             }
         }
+
+        [Test]
+        [IssueLink("https://github.com/npgsql/npgsql/issues/4099")]
+        public async Task Bug4099()
+        {
+            var csb = new NpgsqlConnectionStringBuilder(ConnectionString)
+            {
+                Multiplexing = true,
+                MaxPoolSize = 1
+            };
+            await using var postmaster = PgPostmasterMock.Start(csb.ConnectionString);
+            await using var firstConn = await OpenConnectionAsync(postmaster.ConnectionString);
+            await using var secondConn = await OpenConnectionAsync(postmaster.ConnectionString);
+
+            var byteArrayLength = csb.WriteBufferSize + 100;
+            var firstQuery = firstConn.ExecuteScalarAsync("SELECT data");
+
+            var server = await postmaster.WaitForServerConnection();
+            await server.ExpectExtendedQuery();
+
+            var secondQuery = secondConn.ExecuteScalarAsync("SELECT other_data");
+            await server.ExpectExtendedQuery();
+
+            var data = new byte[10000];
+            await server
+                .WriteParseComplete()
+                .WriteBindComplete()
+                .WriteRowDescription(new FieldDescription(PostgresTypeOIDs.Bytea))
+                .WriteDataRowWithFlush(data);
+
+            var otherData = new byte[10];
+            await server
+                .WriteCommandComplete()
+                .WriteReadyForQuery()
+                .WriteParseComplete()
+                .WriteBindComplete()
+                .WriteRowDescription(new FieldDescription(PostgresTypeOIDs.Bytea))
+                .WriteDataRow(otherData)
+                .WriteCommandComplete()
+                .WriteReadyForQuery()
+                .FlushAsync();
+
+            CollectionAssert.AreEqual(data, (byte[])(await firstQuery)!);
```

Comment:
nit: I think this is that same as `Assert.That(data, Is.EquivalentTo(...))` - we tend to use that syntax rather than the classical one (Assert.Equal etc.). Up to you.